KAFKA-15167: Tiered Storage Test Harness Framework

kamalcph opened the pull request:
`TieredStorageTestHarness` is a base class for integration tests exercising the tiered storage functionality. This uses  `LocalTieredStorage` instance as the second-tier storage system and `TopicBasedRemoteLogMetadataManager` as the remote log metadata manager.

Co-authored-by: Alexandre Dupriez <alexandre.dupriez@gmail.com>
Co-authored-by: Kamal Chandraprakash <kamal.chandraprakash@gmail.com>

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

kamalcph committed:
KAFKA-15167: Tiered Storage Test Harness Framework

Addressed the review comments.

Add the produce and consume actions after calling `produce` or `consume`

- Address the review comments.

Disable tiered storage integration tests.

Fix scala 2.12 compilation issues.

- Minor improvements.
- Update earliestOffset to earliestLocalOffset

Added Offload and consume from leader test.

Added ActionUtils

Fix the checkstyle after rebase with trunk

Fix the checkstyle error by introducing import-control-storage.xml file

Fix compilation issues.

Addressed the remaining TODOs.

TODO items are addressed.

TODO items are addressed.

fix the compilation issues.

Added Record KV matcher

1. Dump local tiered storage
2. Local tiered storage output and
3. Test contexts are added

add broker local storage actions.

Add builder for tiered storage test.

All the action templates are added.

Added more actions.

clolov commented on storage/src/test/java/integration/kafka/tiered/storage/actions/CreateTopicAction.java:
```diff
@@ -0,0 +1,74 @@
+
+    @Override
+    public void doExecute(TieredStorageTestContext context) {
+//
```
```suggestion
        //
```

clolov commented:
Hello and thank you for starting on this 😊! What is your plan when it comes to reviewing this code - will you be splitting it into multiple pull requests for individual actions/specifications + tests which use said actions? Otherwise I believe this pull request will fairly quickly become very large to review properly.

kamalcph commented:
> Hello and thank you for starting on this 😊! What is your plan when it comes to reviewing this code - will you be splitting it into multiple pull requests for individual actions/specifications + tests which use said actions? Otherwise I believe this pull request will fairly quickly become very large to review properly.

Thanks for checking on this! Agree that we have to split the PR into multiple for smooth review. Working on the remaining parts of the code to ensure that all the components are working as expected. Once ready, will push them to this PR to have complete view.

showuon commented:
Please let me know when ready for review. Thanks for the work!

kamalcph commented:
@clolov @divijvaidya @showuon @Hangleton @satishd 

1. This PR contains huge changes but most of them are new POJO classes and specific actions. My suggestion is to review all the changes together, then we can divide this PR into multiple once we agree on the approach taken in this PR to do the integration test.
2. Tiered Storage Integration Test Framework depends upon the `core` module kafka packages, so added a new `import-control-storage.xml` which allows to import `kafka` packages . We can refactor the code (package names) once it gets finalised. 
3. Increased the `ClassDataAbstractionCoupling` from 25 to 40 in `checkstyle.xml` to unblock for now. Once the changes are finalised, the `TieredStorageTestBuilder` class can be further divided to address the checkstyle error.
4. Wrote one integration `OffloadAndConsumeFromLeaderTest` test which produces records, expects the segment to be copied to remote storage and asserts that the consumer able to consume from tiered storage. Right now, `trunk` build is not stable and this test is in Failed test. We can use this framework as TDD to fix issues in the source code.
5. Java does not supports named parameter arguments so the `OffloadAndConsumeFromLeaderTest` is not readable. We can improve the readability aspects in the next PR.

Please take a look when you get chance!

divijvaidya commented on storage/src/test/java/org/apache/kafka/tiered/storage/integration/OffloadAndConsumeFromLeaderTest.java:
```diff
@@ -0,0 +1,136 @@
+ * Test Cases (A):
+ *    Elementary offloads and fetches from tiered storage.
+ */
+public final class OffloadAndConsumeFromLeaderTest extends TieredStorageTestHarness {
```
Since this test is currently failing, I will suggest to mark it as `@disabled` since it will add noise for other PRs and they may wonder why this test is failing. Folks who are working on TS related changes can choose to remove this once it passes.

kamalcph replied:
Disabled the `TieredStorageTestHarness` test until the trunk build becomes stable to test them out.

divijvaidya commented on core/src/main/scala/kafka/server/KafkaBroker.scala:
```diff
@@ -19,6 +19,7 @@ package kafka.server
 
 import com.yammer.metrics.core.MetricName
 import kafka.log.LogManager
+import kafka.log.remote.RemoteLogManager
```
Extra import?

kamalcph replied:
This import is being used in L83.

Removed the unused import "scala.collection.Seq" from this class, but it throwing below error, so reverted it back since the change is unrelated to this PR:

```
❯ ./gradlew :storage:compileTestJava -PscalaVersion=2.13

> Configure project :
Starting build with version 3.6.0-SNAPSHOT (commit id d109f88d) using Gradle 8.2.1, Java 11 and Scala 2.13.11
Build properties: maxParallelForks=12, maxScalacThreads=8, maxTestRetries=0

> Task :core:compileScala
[Error] /Users/kchandraprakash/projects/kafka/core/src/main/scala/kafka/server/DynamicBrokerConfig.scala:910:69: type mismatch;
 found   : scala.collection.mutable.Buffer[org.apache.kafka.common.metrics.MetricsReporter]
 required: Seq[AnyRef]
[Error] /Users/kchandraprakash/projects/kafka/core/src/main/scala/kafka/server/DynamicBrokerConfig.scala:915:61: type mismatch;
 found   : scala.collection.mutable.Buffer[org.apache.kafka.common.metrics.MetricsReporter]
 required: Seq[AnyRef]
[Error] /Users/kchandraprakash/projects/kafka/core/src/main/scala/kafka/server/KafkaServer.scala:267:77: type mismatch;
 found   : Seq[Object] (in scala.collection) 
 required: Seq[AnyRef] (in scala.collection.immutable) 
[Error] /Users/kchandraprakash/projects/kafka/core/src/main/scala/kafka/server/KafkaServer.scala:117:15: private val kafkaMetricsReporters in class KafkaServer is never used
four errors found
```

divijvaidya replied:
haha this is interesting. Let's keep it. I am good as long as build is fine.

divijvaidya commented on storage/src/test/java/org/apache/kafka/tiered/storage/integration/OffloadAndConsumeFromLeaderTest.java:
```diff
@@ -0,0 +1,136 @@
+                 *       - For topic B, only one segment is present in the tiered storage, as asserted by the
+                 *         previous sub-test-case.
+                 */
+                //.bounce(broker)
```
Is this intentional?

I assume we will need it to replicate the "Stops and restarts the broker" scenario.

divijvaidya commented on storage/src/test/java/org/apache/kafka/tiered/storage/TieredStorageTestHarness.java:
```diff
@@ -0,0 +1,205 @@
+        return JavaConverters.asScalaBuffer(configs).toSeq();
+    }
+
+    public Properties getOverridingProps() {
```
in Kafka code base, we don't use prefix such as "get". May I suggest renaming this to overrideProps()

divijvaidya commented on storage/src/test/java/org/apache/kafka/tiered/storage/TieredStorageTestHarness.java:
```diff
@@ -0,0 +1,205 @@
+        // You can override this property with a fixed path of your choice if you wish to use a non-temporary
+        // directory to access its content after a test terminated.
+        overridingProps.setProperty(storageConfigPrefix(STORAGE_DIR_CONFIG), TestUtils.tempDir().getAbsolutePath());
+        overridingProps.setProperty(KafkaConfig.UncleanLeaderElectionEnableProp(), "true");
```
this should default to false to match the default value of Kafka 3.6.0. Is there a reason we are setting this as true here?

divijvaidya commented on storage/src/test/java/org/apache/kafka/tiered/storage/TieredStorageTestHarness.java:
```diff
@@ -0,0 +1,205 @@
+    }
+
+    @Test
+    public void executeTieredStorageTest() {
```
Please add a tag @Tag("integration") which tells JUnit5 that any class extending this class will be run as part of ./gradlew integrationTest and not with ./gradlew unitTest

divijvaidya commented on storage/src/test/java/org/apache/kafka/tiered/storage/TieredStorageTestHarness.java:
```diff
@@ -0,0 +1,205 @@
+        return storages;
+    }
+
+    @SuppressWarnings("deprecation")
```
what is the deprecated thing that we are using here and why?

(same question for rest of the places)

kamalcph replied:
`JavaConverters` is deprecated and `CollectionConverters` will work only from scala 2.13. To make the source code compatible to compile in scala 2.12, used the `JavaConverters` and suppressed the deprecation warnings.

divijvaidya commented on core/src/test/scala/unit/kafka/utils/TestUtils.scala:
```diff
@@ -216,6 +216,14 @@ object TestUtils extends Logging {
+  /**
+   * Create a test config for the provided parameters.
+   */
+  def createServerConfigs(numConfigs: Int,
```
why do we need to add this? Can't we directly use `createBrokerConfigs`?

kamalcph replied:
Removed this method. With `createBrokerConfigs`, we had to supply 18 parameters from Java code.

divijvaidya commented on storage/src/test/java/org/apache/kafka/tiered/storage/TieredStorageTestHarness.java:
```diff
@@ -0,0 +1,205 @@
+
+    @SuppressWarnings("deprecation")
+    @Override
+    public Seq<KafkaConfig> generateConfigs() {
```
I don't think we need to override the entire method. IntegrationTestHarness provides an alternative way to override configs using `modifyConfigs()` method. You can see how it could be used in other examples such as: https://github.com/apache/kafka/blob/7a2e11cae739f3391f61e2e29b148d3a3ebea8b3/core/src/test/scala/unit/kafka/server/BaseRequestTest.scala#L46

The advantage of using `modifyConfigs` instead of current method is that we don't end up losing additional configuration that `IntegrationTestHarness#generateConfigs` does such as automatically adding:
```
    if (isZkMigrationTest()) {
      cfgs.foreach(_.setProperty(KafkaConfig.MigrationEnabledProp, "true"))
    }
```

divijvaidya commented on storage/src/test/java/org/apache/kafka/tiered/storage/TieredStorageTestHarness.java:
```diff
@@ -0,0 +1,205 @@
+
+    private TieredStorageTestContext context;
+
+    protected int numMetadataPartition() {
```
this is a very confusing name since it could mean partitions for consumer metadata topic or kraft metadata topic. can you please rename this appropriately. Also, could this be moved to a member in this class instead (which can be accessed by subClasses).

divijvaidya commented on storage/src/test/java/org/apache/kafka/tiered/storage/TieredStorageTestContext.java:
```diff
@@ -0,0 +1,348 @@
+        // Set a producer linger of 60 seconds, in order to optimistically generate batches of
+        // records with a pre-determined size.
+        producerConfig.put(LINGER_MS_CONFIG, String.valueOf(TimeUnit.SECONDS.toMillis(60)));
+        producer = new KafkaProducer<>(producerConfig, ser, ser);
```
please use `IntegrationTestHarness.createProducer()` method to initialize the producer in `TieredStorageTestHarness`

kamalcph replied:
IntegrationTestHarness.createProducer() -- creates a new producer instance on each invocation.

divijvaidya commented on storage/src/test/java/org/apache/kafka/tiered/storage/TieredStorageTestContext.java:
```diff
@@ -0,0 +1,348 @@
+        producer = new KafkaProducer<>(producerConfig, ser, ser);
+
+        consumerConfig.put(BOOTSTRAP_SERVERS_CONFIG, bootstrapServers);
+        consumer = new KafkaConsumer<>(consumerConfig, de, de);
```
please use `IntegrationTestHarness.createConsumer()` method to initialize the producer in `TieredStorageTestHarness`

kamalcph replied:
IntegrationTestHarness.createConsumer() -- creates a new consumer instance and closes it on each invocation.

divijvaidya commented on storage/src/test/java/org/apache/kafka/tiered/storage/TieredStorageTestContext.java:
```diff
@@ -0,0 +1,348 @@
+
+        Properties adminConfig = new Properties();
+        adminConfig.put(BOOTSTRAP_SERVERS_CONFIG, bootstrapServers);
+        admin = Admin.create(adminConfig);
```
please use `IntegrationTestHarness.createAdminClient()` method to initialize the producer in `TieredStorageTestHarness`

kamalcph replied:
same as above -- creates a new instance for each invocation.

divijvaidya commented on storage/src/test/java/org/apache/kafka/tiered/storage/TieredStorageTestContext.java:
```diff
@@ -0,0 +1,348 @@
+    }
+
+    private void initContext() {
+        String bootstrapServers = TestUtils.bootstrapServers(brokers, listenerName);
```
we don't have to do this manually. IntegrationTestHarness does it for you. When you use TieredStorageHarness, you will automatically have brokers initialized.

divijvaidya commented on storage/src/test/java/org/apache/kafka/tiered/storage/TieredStorageTestContext.java:
```diff
@@ -0,0 +1,348 @@
+            newTopic = new NewTopic(spec.getTopicName(), replicasAssignments);
+        }
+        newTopic.configs(spec.getProperties());
+        admin.createTopics(Collections.singletonList(newTopic)).all().get();
```
please use `KafkaServerTestHarness.createTopic()` which is available in `TieredStorageTestHarness` since it's a super class of `IntegrationTestHarness `

kamalcph replied:
same as above -- creates the Admin instance on each invocation. 

To write complex tests, we need the same instance for multiple actions. Creating it for each action might be expensive and difficult to read through the logs.

divijvaidya replied:
this create topic execution will not wait until topic metadata information has been propagated to all broker which could lead to flaky test. I understand we want to re-use admin client. In that case, could we use `TestUtils.createTopicWithAdmin` instead?

divijvaidya replied:
@kamalcph planning to address this comment?

kamalcph replied:
Used `TestUtils.waitForAllPartitionsMetadata` instead of `TestUtils.createTopicWithAdmin` as the latter requires converting the replicaAssignment Map from java to scala. Will do it later once the PR is stable.

divijvaidya commented on storage/src/test/java/org/apache/kafka/tiered/storage/TieredStorageTestContext.java:
```diff
@@ -0,0 +1,348 @@
+import static org.apache.kafka.clients.CommonClientConfigs.BOOTSTRAP_SERVERS_CONFIG;
+import static org.apache.kafka.clients.producer.ProducerConfig.LINGER_MS_CONFIG;
+
+public final class TieredStorageTestContext {
```
I think that a lot of the functionality in this class already exists in `IntegrationTestHarness` or it's parent classes. I left comments for few examples but I think we can simply re-use the existing methods and for the ones which we cannot re-use, we can fold them into Kafka `TieredStorageTestHarness`. We can re-factor this class out later if required from `TieredStorageTestHarness`.

kamalcph replied:
`IntegrationTestHarness` creates instance for each action invocation. We can do the refactoring later if required.

divijvaidya replied:
We can still use a lot of functions from `IntegrationTestHarness` such as `createAdminClient` to create the admin client for the first time. I think it's important to do this right now else it may lead to resource leak.

Also, we should be be using TestUtils for various methods here since the TestUtils ensure that metadata has propagated to the brokers correctly before any assertions are made later.

divijvaidya commented on storage/src/test/java/org/apache/kafka/tiered/storage/TieredStorageTestContext.java:
```diff
@@ -0,0 +1,348 @@
+        admin.incrementalAlterConfigs(configsMap, alterOptions).all().get(30, TimeUnit.SECONDS);
+    }
+
+    public void deleteTopic(String topic) throws ExecutionException, InterruptedException, TimeoutException {
```
again, `KafkaServerTestHarness.deleteTopic()` could be used directly here. We don't need to implement this.

kamalcph replied:
ditto

divijvaidya replied:
Replace with `TestUtils.deleteTopicWithAdmin` in that case if we want to re-use admin client.

divijvaidya commented on storage/src/test/java/org/apache/kafka/tiered/storage/TieredStorageTestBuilder.java:
```diff
@@ -0,0 +1,427 @@
+        return this;
+    }
+
+    public TieredStorageTestBuilder produce(String topic,
```
when I call this method, it doesn't add produce actions but it does add consume actions for any pending consumables. Why? I am not sure I understand why aren't we adding produce actions at the end of this method AND why add consume actions added by some previous calls when invoking this call? Why do we have the concept of producibles or consumables?

kamalcph replied:
Updated the `produce` and `consume` methods to be deterministic. 

Previously, the DSL can be written as:
```
.produce(topicA, p0, new KeyValueSpec("k1", "v1"), new KeyValueSpec("k2", "v2"))
.withBatchSize(topicA, p0, batchSize)
.expectSegmentToBeOffloaded(broker, topicA, p0, 0, new KeyValueSpec("k1", "v1"))
.expectEarliestLocalOffsetInLogDirectory(topicBA, p0, 1L)
```

Now, with this change, the expectation needs to be called before the `produce` action:
```
.withBatchSize(topicA, p0, batchSize)
.expectSegmentToBeOffloaded(broker, topicA, p0, 0, new KeyValueSpec("k1", "v1"))
.expectEarliestLocalOffsetInLogDirectory(topicBA, p0, 1L)
.produce(topicA, p0, new KeyValueSpec("k1", "v1"), new KeyValueSpec("k2", "v2"))
```

Updated the code as the changes are intuitive to the user.

Similarly for the consume method.

divijvaidya replied:
Thank you. It makes much more sense now.

divijvaidya commented on storage/src/test/java/org/apache/kafka/tiered/storage/TieredStorageTestBuilder.java:
```diff
@@ -0,0 +1,427 @@
+        return this;
+    }
+
+    public TieredStorageTestBuilder deleteRecords(String topic,
```
suggestion - (for this and other such methods)

Change to take an optional argument of expected exception.
`def deleteRecords(topic: String, partition: Int, beforeOffset: Int, expectedException: Option[Class[_]] = None)`

and deleteRecordsAction should do:
```
try {
      context.admin().deleteRecords(recordsToDelete).all().get
      expectedException.map(exp => fail("Expected exception: " + exp))
    } catch {
      case e: Throwable => {
        if(expectedException.isDefined) {
          assertEquals(expectedException.get, e.getCause.getClass, "Unexpected exception cause " + e.getCause)
        }
      }
```

kamalcph replied:
Not clear on this one. We are not expecting any exception in the deleteRecords. Could you show me the exact method/line to update?

divijvaidya replied:
I mentioned this as an example of how this interface can help write tests where we want to assert an exception on the client. DeleteRecords was probably not the best choice here but the idea was to demonstrate how we can change the harness to expect exceptions.

We can do this later outside this PR if a need arises. This is optional.

divijvaidya commented on storage/src/test/java/org/apache/kafka/tiered/storage/actions/ProduceAction.java:
```diff
@@ -0,0 +1,163 @@
+            }
+        }
+
+        List<ConsumerRecord<String, String>> consumedRecords =
+                context.consume(topicPartition, recordsToProduce.size(), startOffset);
```
why are we consuming in a produce action? I am assuming that the reason is to validate that produce was executed correctly. yes? Please add a comment about this here.

divijvaidya commented on storage/src/test/java/org/apache/kafka/tiered/storage/integration/OffloadAndConsumeFromLeaderTest.java:
```diff
@@ -0,0 +1,136 @@
+ * Test Cases (A):
+ *    Elementary offloads and fetches from tiered storage.
+ */
+public final class OffloadAndConsumeFromLeaderTest extends TieredStorageTestHarness {
```
we need a lightweight README.md in tiered/storage folder. That README should explain the basic construction of a test.

Step 1: For every test, setup is done via `TieredStorageTestHarness` which extends `IntegrationTestHarness` and sets up a cluster with TS enabled on it.
Step 2: The test is written as a specification consisting of sequential actions and assertions. The spec for the complete test is written down first which creates "actions" to be executed.
Step 3: Once we have the test spec in-place (which includes assertion actions), we execute the test which will execute each action sequentially.
Step 4: The test execution stops when any of the action throws an exception (or an assertion error). 
Step 5: Clean-up for the test is performed on test exit

divijvaidya commented on storage/src/test/java/org/apache/kafka/tiered/storage/TieredStorageTestBuilder.java:
```diff
@@ -0,0 +1,427 @@
+        assert maxBatchCountPerSegment >= 1 : "Segments size for topic " + topic + " needs to be >= 1";
+        assert partitionCount >= 1 : "Partition count for topic " + topic + " needs to be >= 1";
+        assert replicationFactor >= 1 : "Replication factor for topic " + topic + " needs to be >= 1";
+        maybeCreateProduceAction();
+        maybeCreateConsumeActions();
```
addition of these actions is very confusing for me. My understanding is that we add these actions because the parent action is dependent on it? But for example, in this case, createTopicAction is not dependent on createProduce/ConsumeAction, then why do we add them before this?

divijvaidya reviewed (commented):
My main concern is that we are using IntegrationTestHarness but duplicating a lot of it's code in our own implementation in TieredStorageContext. This leads to a situation where the tests we add will not be runnable with kraft/zk mode easily.

Instead of reviewing this and then breaking it down, let's merge this PR in with only the things required for sample integration test to work. I am comfortable reviewing this PR as it is right now (assuming we remove unnecessary specs and actions).

My suggestion:
1. Remove TieredStorageContext and pass instance of TieredStorageTestHarness to `TieredStorageTestAction` so that actions have access to utility methods when actually executing.
2. Remove the concept of producible and consumables. I don't understand how they are adding value. Instead work directly with ProduceActions and ConsumeActions. Actually, consider removing the entire `specs` folder unless you think they are needed.
3. Remove all actions for this PR and only keep the ones required to run the sample integration test. Others actions could be added in follow-up PRs when they are required by a integration test.

kamalcph commented:
> This leads to a situation where the tests we add will not be runnable with kraft/zk mode easily.

We are not using ZooKeeper client directly anywhere so Kraft mode should work as expected.

> My main concern is that we are using IntegrationTestHarness but duplicating a lot of it's code in our own implementation in TieredStorageContext

Agree, there are some code duplications but it gives us flexibility to reuse the client which avoids log pollution while debugging a failed test.

>  unnecessary specs and actions

Plan is to add respective tests such as ReassignReplicaShrinkTest, ReassignReplicaExpandTest, ReassignReplicaMoveTest, PartitionsExpandTest, DeleteSegmentsByRetentionTimeTest, ListOffsetsTest and so on once the build becomes stable. We can remove the unused actions if required.

Thanks for reviewing the PR!

divijvaidya commented on storage/src/test/java/org/apache/kafka/tiered/storage/TieredStorageTestHarness.java:
```diff
@@ -0,0 +1,204 @@
+import static org.apache.kafka.server.log.remote.storage.LocalTieredStorage.STORAGE_DIR_CONFIG;
+
+/**
+ * Base class for integration tests exercising the tiered storage functionality in Apache Kafka.
```
please add that this uses LocalStore as RSM and TopicRLMM as RLMM

divijvaidya commented on storage/src/test/java/org/apache/kafka/tiered/storage/TieredStorageTestHarness.java:
```diff
@@ -0,0 +1,204 @@
+                TopicBasedRemoteLogMetadataManager.class.getName());
+        overridingProps.setProperty(REMOTE_LOG_MANAGER_TASK_INTERVAL_MS_PROP, "1000");
+
+        overridingProps.setProperty(REMOTE_STORAGE_MANAGER_CONFIG_PREFIX_PROP, storageConfigPrefix(""));
```
should we have a separate prefix per test? It will help disambiguate when different tests are running simultaneously. We can use `testInfo.getDisplayName()` as a prefix.

(same for RSM prefix)

kamalcph replied:
Used the test class name in the prefix instead of display name as it always be "executeTieredStorageTest".

divijvaidya commented on storage/src/test/java/org/apache/kafka/tiered/storage/TieredStorageTestHarness.java:
```diff
@@ -0,0 +1,204 @@
+        overridingProps.setProperty(REMOTE_STORAGE_MANAGER_CLASS_NAME_PROP, LocalTieredStorage.class.getName());
+        overridingProps.setProperty(REMOTE_LOG_METADATA_MANAGER_CLASS_NAME_PROP,
+                TopicBasedRemoteLogMetadataManager.class.getName());
+        overridingProps.setProperty(REMOTE_LOG_MANAGER_TASK_INTERVAL_MS_PROP, "1000");
```
please add "1000" to a constant here and explain in javadoc why are we setting it to 1s (because the default is 30s and will test will take forever). Can we reduce it even further to say 500ms to reduce the overall test runtime?

divijvaidya commented on storage/src/test/java/org/apache/kafka/tiered/storage/TieredStorageTestHarness.java:
```diff
@@ -0,0 +1,204 @@
+        // in every broker and throughout the test. Indeed, as brokers are restarted during the test.
+        // You can override this property with a fixed path of your choice if you wish to use a non-temporary
+        // directory to access its content after a test terminated.
+        overridingProps.setProperty(storageConfigPrefix(STORAGE_DIR_CONFIG), TestUtils.tempDir().getAbsolutePath());
```
The path should be unique per-test and per-broker since multiple tests may run simultaneously and access the same directory. May I suggest - `TestUtils.tempDirectory(s"kafka-remotetier-${this.getClass.getSimpleName}"`

divijvaidya commented on storage/src/test/java/org/apache/kafka/tiered/storage/TieredStorageTestHarness.java:
```diff
@@ -0,0 +1,204 @@
+    public static List<LocalTieredStorage> getTieredStorages(Seq<KafkaBroker> brokers) {
+        List<LocalTieredStorage> storages = new ArrayList<>();
+        JavaConverters.seqAsJavaList(brokers).forEach(broker -> {
+            if (broker.remoteLogManagerOpt().isDefined()) {
```
else throw error with a message.

We never expect system level logManager not to be enabled in these tests when this function is called.

divijvaidya commented on storage/src/test/java/org/apache/kafka/tiered/storage/TieredStorageTestReport.java:
```diff
@@ -0,0 +1,76 @@
+            }
+        }
+        String lts = "";
+        if (!context.getTieredStorages().isEmpty()) {
```
is it possible for this to be empty if the test is set up correctly? Isn't it an error case if no tiered storage is configured?

kamalcph replied:
If the test is setup correctly, then the `getTieredStorages` (or) `remoteStorageManagers` won't be empty. Added the `isEmpty` check to avoid NPE. We can remove/refactor the code later.

divijvaidya commented on storage/src/test/java/org/apache/kafka/tiered/storage/TieredStorageTestReport.java:
```diff
@@ -0,0 +1,76 @@
+    }
+
+    public void addSucceeded(TieredStorageTestAction action) {
+        synchronized (this) {
```
synchronized(this) is equivalent to adding synchronized on the method.

divijvaidya commented on storage/src/test/java/org/apache/kafka/tiered/storage/utils/BrokerLocalStorage.java:
```diff
@@ -0,0 +1,162 @@
+                            "in the log directory is %d which is %s the expected offset %s. The directory of %s is " +
+                            "made of the following files: %s", brokerId, topicPartition,
+                    offsetHolder.firstLogFileBaseOffset, pos, offset, topicPartition,
+                    Utils.join(offsetHolder.partitionFiles, "\n"));
```
please replace usage of "\n" with `System.lineSeparator()`

(here and if used elsewhere)

divijvaidya commented on storage/src/test/java/org/apache/kafka/tiered/storage/TieredStorageTestHarness.java:
```diff
@@ -0,0 +1,204 @@
+public abstract class TieredStorageTestHarness extends IntegrationTestHarness {
+
+    /**
+     * InitialTaskDelayMs is set to 30 seconds for the delete-segment scheduler in Apache Kafka.
```
should we add an assertion in the setup to ensure that STORAGE_WAIT_TIMEOUT_SEC > REMOTE_LOG_MANAGER_TASK_INTERVAL_MS_PROP? You can potentially add this assertion in `overridingProps()` below

divijvaidya commented on storage/src/test/java/org/apache/kafka/tiered/storage/TieredStorageTestHarness.java:
```diff
@@ -0,0 +1,204 @@
+    @Override
+    public void tearDown() {
+        try {
+            context.close();
```
`Utils.closeQuietly()`, otherwise on an exception during close we may not end up calling super.teardown() which will leak resources.

divijvaidya commented on storage/src/test/java/org/apache/kafka/tiered/storage/TieredStorageTestContext.java:
```diff
@@ -0,0 +1,339 @@
+            newPartitions = NewPartitions.increaseTo(spec.getPartitionCount(), newAssignments);
+        }
+        Map<String, NewPartitions> partitionsMap = Collections.singletonMap(spec.getTopicName(), newPartitions);
+        admin.createPartitions(partitionsMap).all().get();
```
please wait for metadata propagation to all brokers here. TestUtils doesn't have a method for this but I will encourage you to add one there similar to createTopic and deleteTopic.

To wait for propagation, you can add
```
     TestUtils.computeUntilTrue(TestUtils.describeTopic(topicName, admin))(td => {
        td.isDefined && td.get.partitions().asScala.size == newPartitionCount
      })
```

kamalcph replied:
We can reuse the below method for partition expansion:

```
TestUtils.waitForAllPartitionsMetadata(harness.brokers(), spec.getTopicName(), spec.getPartitionCount());
```

divijvaidya commented on storage/src/test/java/org/apache/kafka/tiered/storage/TieredStorageTestContext.java:
```diff
@@ -0,0 +1,339 @@
+        AlterConfigsOptions alterOptions = new AlterConfigsOptions().timeoutMs(30000);
+        Map<ConfigResource, Collection<AlterConfigOp>> configsMap =
+                Collections.singletonMap(configResource, alterEntries);
+        admin.incrementalAlterConfigs(configsMap, alterOptions).all().get(30, TimeUnit.SECONDS);
```
wait for propagation with something similar to:
```
brokers.foreach(broker => TestUtils.waitUntilTrue(() => broker.logManager.producerStateManagerConfig.producerIdExpirationMs == 100000, "Configuration was not updated."))
```

divijvaidya commented on storage/src/test/java/org/apache/kafka/tiered/storage/TieredStorageTestContext.java:
```diff
@@ -0,0 +1,339 @@
+        KafkaBroker broker = brokers.apply(brokerId);
+        closeClients();
+        broker.shutdown();
+        broker.awaitShutdown();
```
I will suggest to use `KafkaServerTestHarness.killBroker` otherwise the information in `alive` var for this test is inconsistent.

divijvaidya commented on storage/src/test/java/org/apache/kafka/tiered/storage/TieredStorageTestContext.java:
```diff
@@ -0,0 +1,339 @@
+        producerConfig.put(LINGER_MS_CONFIG, String.valueOf(TimeUnit.SECONDS.toMillis(60)));
+        producer = new KafkaProducer<>(producerConfig, ser, ser);
+        consumer = new KafkaConsumer<>(consumerConfig, de, de);
+        admin = Admin.create(adminConfig);
```
`IntegrationTestHarness.createAdminClient` please. It will ensure that the admin client is tracked at `adminClients` global variable and cleaned properly.

divijvaidya commented on storage/src/test/java/org/apache/kafka/tiered/storage/TieredStorageTestContext.java:
```diff
@@ -0,0 +1,339 @@
+        return admin;
+    }
+
+    public boolean isActive(Integer brokerId) {
```
KafkaServerTestHarness maintains "alive" brokers. Please use that, KafkaServerTestHarness is the one that is initialising the brokers in the first place and setting up the test.

divijvaidya commented on storage/src/test/java/org/apache/kafka/tiered/storage/TieredStorageTestHarness.java:
```diff
@@ -0,0 +1,204 @@
+    }
+
+    @SuppressWarnings("deprecation")
+    public static List<LocalTieredStorage> getTieredStorages(Seq<KafkaBroker> brokers) {
```
getTieredStorages is confusing here since this method gives an instance of RSM plugins for each broker. Could we rename this to `remoteStorageManagers()`?

divijvaidya commented on storage/src/test/java/org/apache/kafka/tiered/storage/actions/ProduceAction.java:
```diff
@@ -0,0 +1,164 @@
+                                List<ProducerRecord<String, String>> producerRecords,
+                                TopicPartition topicPartition) {
+        assertThat(discoveredRecords, correspondTo(producerRecords, topicPartition, serde, serde));
+//        assertEquals(spec.getBaseOffset(), discoveredRecords.get(0).offset(),
+//                "Base offset of segment mismatch");
```
intentional?

divijvaidya commented on storage/src/test/java/org/apache/kafka/tiered/storage/TieredStorageTestContext.java:
```diff
@@ -0,0 +1,339 @@
+    }
+
+    public LocalTieredStorageSnapshot takeTieredStorageSnapshot() {
+        return LocalTieredStorageSnapshot.takeSnapshot(tieredStorages.get(0));
```
I assume that we want this snapshot from broker(0) because the RemoteStorage data will remain same no matter which RSM you call from? In that case, I will suggest to change "0" to take a snapshot from "alive" broker. List of "alive" brokers is maintained by KafkaServerTestHarness

divijvaidya commented on storage/src/test/java/org/apache/kafka/tiered/storage/TieredStorageTestContext.java:
```diff
@@ -0,0 +1,339 @@
+        return tieredStorages.get(brokerId).getHistory();
+    }
+
+    public List<LocalTieredStorage> getTieredStorages() {
```
1. please remove "get" prefix from all methods. In Kafka code base we avoid getters/setters in general: https://kafka.apache.org/coding-guide
2. this should be renamed to remoteStorageManagers() since they return instance of RSM plugins for all brokers.

divijvaidya reviewed (commented):
> Agree, there are some code duplications but it gives us flexibility to reuse the client which avoids log pollution while debugging a failed test.

Re-using admin client is fair but please use helper methods in TestUtils as they have additional logic to reduce flaky tests. I have left a few examples but haven't covered all places exhaustively during the review. You can still use TestUtils helpers by providing them with your own `admin` client.

> Plan is to add respective tests such as ReassignReplicaShrinkTest, ReassignReplicaExpandTest, ReassignReplicaMoveTest, PartitionsExpandTest, DeleteSegmentsByRetentionTimeTest, ListOffsetsTest and so on once the build becomes stable. We can remove the unused actions if required.

Let's add these tests and actions in separate PRs please. In this PR let's focus on having one test and associated actions & harness framework merged in.

satishd reviewed (commented):
>Let's add these tests and actions in separate PRs please. In this PR let's focus on having one test and associated actions & harness framework merged in.

Right @divijvaidya . The plan is to add different test scenarios in the respective PRs instead of adding them in this PR. This PR is to have the required changes for the framework.

kamalcph committed:
Addressed the below subset of review comments.

- Remove `get` prefix from the method names.
- Use platform independent line separator.
- Use Autocloseable to avoid resource leak and close the resources quietly.
- Supply the instance of `harness` to the test context.
- Use KafkaServerTestHarness#killBroker instead of manual shutdown.
- Rename tieredStorages to remoteStorageManagers.
- Removed magic values and defined constants.
- Added assertions to catch invalid configuration.

kamalcph committed:
Use TestUtils and wait for metadata propagation

divijvaidya commented:
@kamalcph changes look good to me. Let's focus on CI build now. Could you please fix the checkstyle failures which are causing build failures?

showuon commented on checkstyle/checkstyle.xml:
```diff
@@ -119,7 +119,7 @@
     <module name="ClassDataAbstractionCoupling">
       <!-- default is 7 -->
-      <property name="max" value="25"/>
+      <property name="max" value="40"/>
```
Could we suppress it when needed instead of changing the global setting?

showuon commented on storage/src/test/java/org/apache/kafka/tiered/storage/README.md:
```diff
@@ -0,0 +1,9 @@
+Step 1: For every test, setup is done via TieredStorageTestHarness which extends IntegrationTestHarness and sets up a cluster with TS enabled on it.
```
Could we add some header/title above this steps? Maybe "the test flow" ?

kamalcph replied:
Will add detailed steps later once we start to enable the tests.

showuon reviewed (commented):
Will continue to review it tomorrow.

kamalcph committed:
Fix the checkstyle

kamalcph committed:
Fix the assertion.

kamalcph committed:
Used the test classname as prefix in the rsm and rlmm config.

kamalcph commented:
@divijvaidya 

Addressed most of your comments and marked them as resolved. Please take a look that those comments are addressed as expected. Thanks!

divijvaidya commented on storage/src/test/java/org/apache/kafka/tiered/storage/TieredStorageTestContext.java:
```diff
@@ -0,0 +1,322 @@
+    }
+
+    public void bounce(int brokerId) {
+        closeClients();
```
why do we need to close the clients when bouncing a server? The clients should be able to reconnect automatically.

(same for stop, start and other operations where we specifically are playing around with a specific broker)

divijvaidya commented on storage/src/test/java/org/apache/kafka/tiered/storage/TieredStorageTestContext.java:
```diff
@@ -0,0 +1,322 @@
+        Utils.closeQuietly(admin, "Admin client");
+    }
+
+    private void closeClients() {
```
if we are creating clients using the harness, we don't have to close them explicitly. Harness takes care of that. It closes them at the end of the test.

divijvaidya commented on storage/src/test/java/org/apache/kafka/tiered/storage/TieredStorageTestContext.java:
```diff
@@ -0,0 +1,322 @@
+        closeClients();
+        harness.killBroker(brokerId);
+        harness.startBroker(brokerId);
+        initContext();
```
why do we create new clients every time on bounce? The clients should be able to recover and reconnect.

divijvaidya commented on storage/src/test/java/org/apache/kafka/tiered/storage/TieredStorageTestContext.java:
```diff
@@ -0,0 +1,322 @@
+    public void stop(int brokerId) {
+        closeClients();
+        harness.killBroker(brokerId);
+        initContext();
```
after stopping a broker, do we need to create clients again? why?

divijvaidya reviewed (approved):
I have left some comments but they can be fixed in separate PR and not blocking. Let's wait for @showuon to review tomorrow as he mentioned above.

kamalcph committed:
use test class name instead of the display name.

kamalcph commented:
@showuon 
Call for review. PTAL.

kamalcph committed:
Don't reinitialize the clients on broker stop, start (or) restart.

showuon commented:
@kamalcph , sorry, I'll attend a conference the following days, so I can't review this until next Tuesday. Does this PR block the following PRs? If so, I think since Divij already approved it, we can merge it first, and we can have a follow-up PRs to address my comments after my review next week. WDYT?

kamalcph commented:
@satishd 

Can you please take a look when you get chance?

kamalcph committed:
Remove the unused imports.

kamalcph commented:
How to fix the below error in the test? The test is currently marked as DISABLED but still fails with below error, there are some left-over threads:

Build / JDK 11 and Scala 2.13 / initializationError – org.apache.kafka.tiered.storage.integration.OffloadAndConsumeFromLeaderTest
```
Suppressed: org.opentest4j.AssertionFailedError: Found 2 unexpected threads during @AfterAll: `controller-event-thread,Test worker-EventThread` ==> expected: <true> but was: <false>
		at app//org.junit.jupiter.api.AssertionFailureBuilder.build(AssertionFailureBuilder.java:151)
```

clolov commented:
Heya @kamalcph! I have seen this error pop up when server.shutdown hasn't been called in integration tests, I would suggest looking into whether for whatever reason there is something which starts servers but doesn't shut them down properly. I will look through the code to lend a hand.

clolov commented:
Huh, yeah, no idea. I have been trying to replicate it locally and it appears to be doing exactly what you would expect it to.

satishd reviewed (approved):
Thanks @kamalcph for the PR and addressing the review comments. LGTM.

satishd commented:
There are a few failed tests that are unrelated to the changes in the PR. 

This PR is blocking to add of several integration tests, merging the changes to trunk.

It received 2 +1s, any remaining comments that need to be taken care of can be done in a followup PR.

showuon commented on storage/src/test/java/org/apache/kafka/tiered/storage/actions/CreateTopicAction.java:
```diff
@@ -0,0 +1,61 @@
+    public void doExecute(TieredStorageTestContext context) throws ExecutionException, InterruptedException {
+        // Ensure offset and time indexes are generated for every record.
+        spec.getProperties().put(TopicConfig.INDEX_INTERVAL_BYTES_CONFIG, "1");
+        // Leverage the use of the segment index size to create a log-segment accepting one and only one record.
+        // The minimum size of the indexes is that of an entry, which is 8 for the offset index and 12 for the
+        // time index. Hence, since the topic is configured to generate index entries for every record with, for
+        // a "small" number of records (i.e. such that the average record size times the number of records is
+        // much less than the segment size), the number of records which hold in a segment is the multiple of 12
+        // defined below.
+        if (spec.getMaxBatchCountPerSegment() != -1) {
+            spec.getProperties().put(
+                    TopicConfig.SEGMENT_INDEX_BYTES_CONFIG, String.valueOf(12 * spec.getMaxBatchCountPerSegment()));
```
Could you let me know why it is not `8 * spec.getMaxBatchCountPerSegment()` for offset index?

kamalcph replied:
This is the allocated size for all the indexes. Since, timeIndex requires 12 bytes for one record batch which is higher than the offsetIndex of 8 bytes.

showuon commented on storage/src/test/java/org/apache/kafka/tiered/storage/actions/ShrinkReplicaAction.java:
```diff
@@ -0,0 +1,83 @@
+
+import static org.apache.kafka.tiered.storage.utils.ActionUtils.describeTopic;
+
+public final class ShrinkReplicaAction implements TieredStorageTestAction {
```
Is this a real use case? TBH, I don't know Kafka supports shirnk replica action.

kamalcph replied:
Kafka does not support decreasing the number of partitions but shrinking the replication factor is supported via `ALTER_PARTITION_REASSIGNMENTS` API.

showuon commented on storage/src/test/java/org/apache/kafka/tiered/storage/utils/BrokerLocalStorage.java:
```diff
@@ -0,0 +1,162 @@
+
+    /**
+     * Wait until the first segment offset in Apache Kafka storage for the given topic-partition is
+     * equal or greater to the provided offset.
```
I think we're waiting for `equal` only here, not greater, right?

kamalcph replied:
ack, typo error. Will update the comment in the next PR.

showuon reviewed (commented):
Finally had a pass, left some comments. Thanks again for the effort to put all these test framework!

kamalcph commented:
> Finally had a pass, left some comments. Thanks again for the effort to put all these test framework!

Credits to @Hangleton for coming up with this framework!
